pkg/v1/validate/image.go

```diff
@@ -0,0 +1,276 @@
+// Copyright 2018 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validate
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"io/ioutil"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-containerregistry/pkg/v1"
+)
+
+// Image valides that img does not violate any invariants on the image format.
+func Image(img v1.Image) error {
+	if err := validateLayers(img); err != nil {
+		return fmt.Errorf("validating layers: %v", err)
+	}
+
+	if err := validateConfig(img); err != nil {
+		return fmt.Errorf("validating config: %v", err)
+	}
+
+	if err := validateManifest(img); err != nil {
+		return fmt.Errorf("validating manifest: %v", err)
+	}
+
+	return nil
+}
+
+func validateConfig(img v1.Image) error {
+	cn, err := img.ConfigName()
+	if err != nil {
+		return err
+	}
+
+	rc, err := img.RawConfigFile()
+	if err != nil {
+		return err
+	}
+
+	hash, size, err := v1.SHA256(bytes.NewReader(rc))
+	if err != nil {
+		return err
+	}
+
+	m, err := img.Manifest()
+	if err != nil {
+		return err
+	}
+
+	cf, err := img.ConfigFile()
+	if err != nil {
+		return err
+	}
+
+	pcf, err := v1.ParseConfigFile(bytes.NewReader(rc))
+	if err != nil {
+		return err
+	}
+
+	if cn != hash {
+		return fmt.Errorf("mismatched config digest: ConfigName()=%s, SHA256(RawConfigFile())=%s", cn, hash)
+	}
+
+	if want, got := m.Config.Size, size; want != got {
+		return fmt.Errorf("mismatched config size: Manifest.Config.Size()=%d, len(RawConfigFile())=%d", want, got)
+	}
+
+	if diff := cmp.Diff(pcf, cf); diff != "" {
+		return fmt.Errorf("mismatched config content: (-ParseConfigFile(RawConfigFile()) +ConfigFile()) %s", diff)
+	}
+
+	if cf.RootFS.Type != "layers" {
+		return fmt.Errorf("invalid ConfigFile.RootFS.Type: %q != %q", cf.RootFS.Type, "layers")
+	}
+
+	return nil
+}
+
+func validateLayers(img v1.Image) error {
+	layers, err := img.Layers()
+	if err != nil {
+		return err
+	}
+
+	digests := []v1.Hash{}
+	diffids := []v1.Hash{}
+	sizes := []int64{}
+	for _, layer := range layers {
+		// TODO: Test layer.Uncompressed.
```

Comment:
Should we do something to prevent some real smart person from consuming `stream.Layer` contents during validation, thus preventing it from being usable after that point?

It might just be a documentation issue. `stream.Layer` could make it clearer that you only get one chance to stream the layer, so it better be on its way to a `remote.Image` etc. if you want to do anything useful.

This makes me think we might want a "streaming-validating" image, which only validates things on-demand as the image is read. `validate.Image` could just return a `v1.Image` implementation that returns errors when it sees something that's invalid.

_That_ makes me wonder why validation is opt-in at all, and whether we should just always validate images for sanity all the time. We might break some clients, but only those that were producing invalid images, so 🤷‍♂️

Reply:
Sounds hard 🤷‍♂️